### ISSUE TITLE
[New build - EXPERIMENTAL] 20w36c: Copy-Paste Tool try-out.

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#2881]
So following a very popular demand and discussion with the community this build introduce a basic copy-paste.
The goal for now is to measure the performance impact of such a tool.
With the copy-paste it become possible to cram a massive amount of detail in a relative short time, which would in turn impact performance massively.

This is still wip:
- Command settings and children entity wont be copied.
- Not really made for decal at the moment.
- Building pivot might prove not convenient.

To use the copy-paste:
- select the new Copy-paste tool
- Hold Lect Mouse Button to drag a selection area
- Press the Rotate Key while dragging to scale vertically.
- Then you can build the clipboard content just like a block.
- Hold Right Mouse Button to clear the clipboard.

If the performance test are conclusive the previous point will be supported and the tool will be improved (instance, prefab).

Bugfixes
#2863 Delete a moving mechanism issue
#2840 White painted laser projectile are red

**Comment by tsunamayo** [comment]
Build is up! Now you can break my game ;)

**Comment by Servomotor** [comment]
RIP StEVO

**Comment by BrickMacklin** [comment]
Right now it they can clip, can that be taken out?

**Comment by tsunamayo** [comment]
BrickMacklin => A few things. Is it intended to not copy the colors and stickers at this time? Can we have a way to orient it on more axis? Right now I need to build a box around what I want to copy so it can be oriented the right way.
Brick paint should work, not voxel paint though.
Orientation is a tricky one, start dragging by what you want to be the base.
I think the best would be to re-introduce the base axis switch, for all bricks.

Also what do you mean it can clip? There is a collision check.
Thanks

**Comment by BrickMacklin** [comment]
I have been able to place a copied entity clipped through other blocks.

Both blocks and bricks are placed pasted in their default grey,

I can make it oriented right if I build some blocks or bricks around what I want to copy but more rotate options would make it easier. I would definitely welcome a base axis switch overall.

For its first iteration it is very good. Has not broken the game and for that we thank you.

**Comment by MrNobodyX3** [comment]
#2883

**Comment by tsunamayo** [comment]
@BrickMacklin @MrNobodyX3 I think what could great is that you show me what you want to copy, the copy logic and axis placement it is definitely tricky.

**Comment by Monstertruck-swe** [comment]
There are more pros than cons for the overall building aspect of this tool. Supporting it whenever your time isn't on more important tasks will surely be appreciated in the long run.

**Comment by BrickMacklin** [comment]
@tsunamayo Here is a video example. Notice I put some bricks around the canister so I could copy it in the orientation I needed. But if I tried a different way there was no way to put it right side up

https://youtu.be/nuzKkppi-hg

**Comment by tsunamayo** [comment]
So you want to copy this from the side? It is weird the natural base is from the ground, like it was built... 
I am not sure what you want it to do? It is working as expected in the video, but I guess you expect something different ;)

**Comment by BrickMacklin** [comment]
I am wondering if we could use a one of our keys to rotate the selection in different orientations. On the second selection I could only rotate it on its side. So it would basically be rotated in circles. Could we rotate it on its y-axis as well so I could orient it right side up as it was intended to be? Just in case any of use start a copy selection from an orientation we didn't mean to in the first place

**Comment by Whackjob** [comment]
Thank you, Tsunamayo.  Today I retire my current ship as a relic from the time before cut and paste.  And today I start work on her sister ship.   Thank you for hearing us.   You know how much this meant to us.

**Comment by tsunamayo** [comment]
@Whackjob lol it is just a test for now, and still very much wip, but I need guy like to stress the game and see what happens!

@BrickMacklin you mean during pasting? It works the same as building a normal block for now. If you want a different axis plane you have to drag your selecting in a different fashion like you did. You dont need an additional brick, you can simply start dragging your selection by pointing directly on to your object.

**Comment by Whackjob** [comment]
Already seeing visual bugs with making a weapon on the stand.  Made a new save for the version, doesn't have the carrier in it.  But, might be the blueprint still in memory?  I've deleted the files for my ship and I'm reloading.

**Comment by tsunamayo** [comment]
@Whackjob is it related to the copy-paste?

**Comment by Whackjob** [comment]
No, I don't think it is.  I think the old ship remained in memory somehow.  I've exported the blueprint to steam and deleted it locally and restarted.  The glitch I saw has not remanifested.

**Comment by ultrak2k** [comment]
damn, this is cool, i placed down a good 6th of the bricks from my monstrosity build with copy paste and it's okay on my low end ass laptop
(not because that's how i'd build, just for testing)

**Comment by Whackjob** [comment]
May I make a suggesting as far as pasting goes?  Can you uncouple the placement from the player, and use keys to move the paste area, one brick or block at a time, with the ability to rotate it on all axes?  I'm trying to copy a wide hallway., in example, empty at both ends.  I can't find where I need to "look" at on the main structure to be able to paste it so they match up.

**Comment by Michael-Phoenix** [comment]
> https://github.com/tsunamayo/Starship-EVO/issues/2881#issuecomment-686875261

Having an option for rotating objects on the other axis would be a great addition in any case. Would save hours usually spent on building blocks in weird places, just so you can get the proper rotation of the blocks I actually want to place.

We already have two keys for rotations. Maybe we could repurpose the second one to just rotate objects the second axis? That would be lovely!

**Comment by Michael-Phoenix** [comment]
Rail block "heads" can be copied and pasted, but can't be deleted from the world.
https://youtu.be/H84MYc2njuQ

**Comment by CurioInventorium** [comment]
I would have thought a system whereby you place a copy/paste block onto a structure and then define it's bounds similar to the volumetric sensor. Once the area is chosen, click on the copy button.
Now someone else, either on the same grid or on another grid you put down the copy/paste block. The volume settings are kept if something has being copied to memory. Open its menu and then click on paste.
The pasted blocks will appear in much the same way and process that blocks being mirrored would be pasted. If there's a block in the way it doesn't get placed.
Think of it more being like a script that records the blocks, decals and colours placed within the volume and then just plays back the script when the paste button is pressed.

Buttons to appear in the blocks menu would be Copy, Paste & Reset block.

**Comment by tsunamayo** [comment]
@CurioInventorium not sure to understand, it seems that you have just described what is in the game atm!

**Comment by CurioInventorium** [comment]
@tsunamayo The system implemented uses a tool that the player aims, the suggested system uses a block placed onto an existing grid. I'll see if I can put together a mock-up series of edited screenshots to demonstrate.

**Comment by CurioInventorium** [comment]
@tsunamayo  Forget what I suggested :)
The system you've implemented is way more convenient

**Comment by BrickMacklin** [comment]
@tsunamayo 
![image](https://user-images.githubusercontent.com/37836252/92270758-920beb00-eeb4-11ea-8ec6-dd082e16cb38.png)

Here is an example of a few users that want to take another look at this game. Copy and paste was a big deal for them. I know this is only a test but they are pretty happy about the addition. Thank you for doing this.

**Comment by tiagonugent** [comment]
Hello, @tsunamayo , I'm a composer and I am very interested in your game! I would really love to compose music for it. Is there any way that I can contact you directly? If you would like to hear my work, check out my website: http://tiagonugent.com/index.html
I also composed a demo main theme for the game: https://soundcloud.com/tiagonugent/starship-evo-fan-demo-theme/s-cKmiQ4TO7EC
